aws/resource_aws_lambda_provisioned_concurrency_config.go

```diff
@@ -164,13 +164,21 @@ func resourceAwsLambdaProvisionedConcurrencyConfigDelete(d *schema.ResourceData,
 }
 
 func resourceAwsLambdaProvisionedConcurrencyConfigParseId(id string) (string, string, error) {
-	parts := strings.SplitN(id, ":", 2)
```

Comment:
I was looking into this issue as well. Instead of having this whole function to parse the `id` that is set based on `function_name` and `qualifier`, is there a reason the calling code can't just use `Get("function_name")` and `Get("qualifier")`? Apologies if this is a silly question, as I don't have much experience with Go or TF provider code.

Reply:
Should be pretty easy experiment to run. However, if memory serves from code changes to other TF providers, the only property set on the schema resource at the start of the read operation is `.d.Id()` and application code uses the AWS logical ids encoded in that id to retrieve additional data via the AWS API and set them in the state file (my expectation is that `d.Get("qualifier")` returns nil` at this point in the execution sequence).

To that end, all the necessary AWS logical to perform one or more read requests must be encoded and parsed from the id file. For example, the [aws_iam_role_policy](https://github.com/hashicorp/terraform-provider-aws/blob/bc48c2178fff1c5b1b142481069e45fcb5870bd5/aws/resource_aws_iam_role_policy.go#L146-#L156) splits the id into two components for `roleName` and `policyName` in spite of the fact that both those values also exist as [schema properties.](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/iam_role_policy).

Reply:
Ran into some missing type assertions when I tested this strategy (solvable), but I think these lines add weight to the behavior I noted above:

```go
        # Parse the function name and qualifier from the id
        functionName, qualifier, err := resourceAwsLambdaProvisionedConcurrencyConfigParseId(d.Id())
        ...
        # Make an API call to read the provisioned concurrency config values via the AWS API
        output, err := conn.GetProvisionedConcurrencyConfig(input)
        ...

        # Finally, set the properties in the state file
        # Since `function_name` and `qualifier` are not set until this point
        # it is unlikely that they are readable at an earlier point in the method.
	d.Set("function_name", functionName)
	d.Set("provisioned_concurrent_executions", output.AllocatedProvisionedConcurrentExecutions)
	d.Set("qualifier", qualifier)
```